plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/aggregation/AggregateExpression.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.iceberg.aggregation;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.collect.ImmutableList;
+import io.trino.plugin.iceberg.ColumnIdentity;
+
+import java.util.Objects;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static io.trino.plugin.iceberg.IcebergMetadataColumn.FILE_MODIFIED_TIME;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class AggregateExpression
+{
+    private final String function;
+    private final String argument;
+    private final boolean returnNullOnEmptyGroup;
+    // did not find a cleaner way to find an id for aggregate's synthetic column,
+    // Trino maintains FILE_MODIFIED_TIME as one of the synthetic metadata column.
+    // which is Integer.MAX_VALUE - 1001, so using FILE_MODIFIED_TIME - 1001 to have some buffer.
+    public static final Integer COUNT_AGGREGATE_COLUMN_ID = FILE_MODIFIED_TIME.getId() - 1001;
```

Comment:
we need a central registry for synthetic field IDs we create

however, for expressions (this includes aggregate expressions), we shouldn't use them at all:
we cannot have IDs for every occasion and we will want to distinguish `count(col_a)` and `count(col_b)`, so they cannot have same ID.


My proposal would be to model Aggregation pushdown as a new type of TableHandle, wrapping IcebergTableHandle.

- It would define aggregations and groupings
- it wouldn't require squeezing aggregate expression into IcebergColumnHandle handle concept, so we wouldn't have the fake column ID problem
- it would require explicit handling in the `applyFilter` and `applyProjection` methods, but that's actually a good thing.  Today, the lack of changes in `applyFilter` leads to errors, when applyFilter happens after applyProjection. Please add a test case like below


```
trino:tpch> SELECT count(*)
         -> FROM nation_p
         -> HAVING count(*) > 2;
Query 20231016_123814_00032_65na2 failed: Cannot find field 'count(*)' in struct: struct<1: nationkey: optional long, 2: name: optional string, 3: regionkey: optional long, 4: comment: optional string>
org.apache.iceberg.exceptions.ValidationException: Cannot find field 'count(*)' in struct: struct<1: nationkey: optional long, 2: name: optional string, 3: regionkey: optional long, 4: comment: optional string>
	at org.apache.iceberg.exceptions.ValidationException.check(ValidationException.java:49)
	at org.apache.iceberg.expressions.NamedReference.bind(NamedReference.java:45)
	at org.apache.iceberg.expressions.NamedReference.bind(NamedReference.java:26)
	at org.apache.iceberg.expressions.UnboundPredicate.bind(UnboundPredicate.java:111)
	at org.apache.iceberg.expressions.Projections$BaseProjectionEvaluator.predicate(Projections.java:181)
	at org.apache.iceberg.expressions.Projections$BaseProjectionEvaluator.predicate(Projections.java:135)
	at org.apache.iceberg.expressions.ExpressionVisitors.visit(ExpressionVisitors.java:347)
	at org.apache.iceberg.expressions.Projections$BaseProjectionEvaluator.project(Projections.java:151)
	at org.apache.iceberg.ManifestGroup.lambda$entries$8(ManifestGroup.java:238)
	at com.github.benmanes.caffeine.cache.LocalLoadingCache.lambda$newMappingFunction$3(LocalLoadingCache.java:183)
	at com.github.benmanes.caffeine.cache.UnboundedLocalCache.lambda$computeIfAbsent$2(UnboundedLocalCache.java:294)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708)
	at com.github.benmanes.caffeine.cache.UnboundedLocalCache.computeIfAbsent(UnboundedLocalCache.java:290)
	at com.github.benmanes.caffeine.cache.LocalCache.computeIfAbsent(LocalCache.java:112)
	at com.github.benmanes.caffeine.cache.LocalLoadingCache.get(LocalLoadingCache.java:58)
	at org.apache.iceberg.ManifestGroup.lambda$entries$9(ManifestGroup.java:258)
	at org.apache.iceberg.io.CloseableIterable$5.shouldKeep(CloseableIterable.java:139)
	at org.apache.iceberg.io.FilterIterator.advance(FilterIterator.java:66)
	at org.apache.iceberg.io.FilterIterator.hasNext(FilterIterator.java:49)
	at org.apache.iceberg.io.FilterIterator.advance(FilterIterator.java:64)
	at org.apache.iceberg.io.FilterIterator.hasNext(FilterIterator.java:49)
	at org.apache.iceberg.io.FilterIterator.advance(FilterIterator.java:64)
	at org.apache.iceberg.io.FilterIterator.hasNext(FilterIterator.java:49)
	at org.apache.iceberg.io.CloseableIterator$3.hasNext(CloseableIterator.java:91)
	at org.apache.iceberg.relocated.com.google.common.collect.TransformedIterator.hasNext(TransformedIterator.java:46)
	at org.apache.iceberg.io.CloseableIterable$ConcatCloseableIterable$ConcatCloseableIterator.hasNext(CloseableIterable.java:247)
	at org.apache.iceberg.relocated.com.google.common.collect.TransformedIterator.hasNext(TransformedIterator.java:46)
	at org.apache.iceberg.relocated.com.google.common.collect.TransformedIterator.hasNext(TransformedIterator.java:46)
	at org.apache.iceberg.relocated.com.google.common.collect.Iterators$ConcatenatedIterator.getTopMetaIterator(Iterators.java:1379)
	at org.apache.iceberg.relocated.com.google.common.collect.Iterators$ConcatenatedIterator.hasNext(Iterators.java:1395)
	at org.apache.iceberg.io.CloseableIterator$1.hasNext(CloseableIterator.java:50)
	at io.trino.plugin.iceberg.aggregation.AggregateSplitSource.isFinished(AggregateSplitSource.java:317)
	at io.trino.plugin.iceberg.aggregation.AggregateSplitSource.getNextBatch(AggregateSplitSource.java:199)
```

- would make `IcebergPageSourceProvider` impl simpler. It already handles aggregations totally differently from non-aggregations

Reply:
@findepi 
yeah ideally -> `we need a central registry for synthetic field IDs we create` 
We are using it synthetic Id [at](https://github.com/trinodb/trino/blob/master/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataColumn.java#L31) as well.

For this PR, I want to avoid supporting `HAVING` clause to start with, as we did not find use-case for it, but looks like missed to validate.

This PR, can restrict to `WHERE` clause (which covers majority of the cases).

I think, we can add another TableHandle in the subsequent series of PRs as it will have other changes to `applyFilter` and `applyProjection` as well.

WDYT.